service/lib/agama/storage/proposal_settings_conversion/to_y2storage.rb

```diff
@@ -0,0 +1,189 @@
+        end
+
+        # @param target [Y2Storage::ProposalSettings]
+        def space_policy_conversion(target)
+          target.space_settings.strategy = :bigger_resize
+
+          actions = case settings.space.policy
+          when :delete
+            all_devices.each_with_object({}) { |d, a| a[d] = :force_delete }
+          when :resize
+            all_devices.each_with_object({}) { |d, a| a[d] = :resize }
+          when :keep
+            {}
+          when :custom
+            settings.space.actions
+          end
+
+          target.space_settings.actions = actions
+        end
+
+        # @param target [Y2Storage::ProposalSettings]
+        def volumes_conversion(target)
+          volumes = settings.volumes.map { |v| VolumeConversion.to_y2storage(v) }
+          disabled_volumes = missing_volumes.map do |volume|
+            VolumeConversion.to_y2storage(volume).tap { |v| v.proposed = false }
+          end
+
+          target.volumes = volumes + disabled_volumes
+
+          fallbacks_conversion(target)
+        end
+
+        # @return [Array<Agama::Storage::Volume>]
+        def missing_volumes
+          mount_paths = settings.volumes.map(&:mount_path)
+
+          VolumeTemplatesBuilder.new_from_config(config).all
+            .reject { |t| mount_paths.include?(t.mount_path) }
+            .reject { |t| t.mount_path.empty? }
+        end
+
+        # @param target [Y2Storage::ProposalSettings]
+        def fallbacks_conversion(target)
+          target.volumes.each do |spec|
+            spec.fallback_for_min_size = find_min_size_fallback(spec.mount_point)
+            spec.fallback_for_max_size = find_max_size_fallback(spec.mount_point)
+          end
+        end
+
+        # @param mount_path [String]
+        def find_min_size_fallback(mount_path)
+          volume = settings.volumes.find { |v| v.min_size_fallback_for.include?(mount_path) }
+          volume&.mount_path
+        end
+
+        # @param mount_path [String]
+        def find_max_size_fallback(mount_path)
+          volume = settings.volumes.find { |v| v.max_size_fallback_for.include?(mount_path) }
+          volume&.mount_path
+        end
+
+        # All block devices affected by the space policy.
+        #
+        # This includes the partitions from the root device, the candidate devices and from the
+        # devices directly assigned to a volume as target device. If a device is not partitioned,
+        # then the device itself is included.
+        #
+        # @return [Array<String>]
+        def all_devices
+          devices = [settings.boot_device] +
+            settings.lvm.system_vg_devices +
+            settings.volumes.map(&:device)
+
+          devices.compact.uniq.map { |d| device_or_partitions(d) }.flatten
+        end
+
+        # @param device [String]
+        # @return [String, Array<String>]
+        def device_or_partitions(device)
+          partitions = partitions(device)
+          partitions.any? ? partitions : device
```

Comment:
can partitions return array with some nils? it does not look for me. So I would maybe use more readable ` partitions.empty? ? device : partitions` to avoid confusion

Reply:
Well, I am very used to `#any?` and `#none?`. For me it is even easier to read than `#empty?`. But let's use `#empty?` if you think it is more common.